### PR TITLE
consensus, core/*, params: metropolis preparation refactor

### PR DESCRIPTION
**Pull request by obscuren**
This commit is a preparation for the upcoming metropolis hardfork. It
prepares the state, core and vm packages such that integration with
metropolis becomes less of a hassle.

* Difficulty calculation requires header instead of individual
  parameters
* statedb.StartRecord renamed to statedb.Prepare and added Finalise
  method required by metropolis, which removes unwanted accounts from
  the state (i.e. selfdestruct)
* State keeps record of destructed objects (in addition to dirty
  objects)
* core/vm pre-compiles may now return errors
* core/vm pre-compiles gas check now take the full byte slice as argument
  instead of just the size
* core/vm now keeps several hard-fork instruction tables instead of a
  single instruction table and removes the need for hard-fork checks in
  the instructions
* core/vm contains a empty restruction function which is added in
  preparation of metropolis write-only mode operations
* Adds the bn256 curve
* Adds and sets the metropolis chain config block parameters (2^64-1)

Follow up PR which adds metropolis features may be found here https://github.com/ethereum/go-ethereum/pull/14337

### CODE REVIEW

File: consensus/ethash/consensus.go [hunk 1]
```diff
@@ -296,15 +299,15 @@ var (
...
 	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.mediawiki
 	// algorithm:
 	// diff = (parent_diff +
 	//         (parent_diff / 2048 * max(1 - (block_timestamp - parent_timestamp) // 10, -99))
 	//        ) + 2^(periodCount - 2)
 
-	bigTime := new(big.Int).SetUint64(time)
-	bigParentTime := new(big.Int).SetUint64(parentTime)
+	bigTime := new(big.Int).Set(parent.Time)
+	bigParentTime := new(big.Int).Set(parent.Time)
```
Comment (holiman) [inline consensus/ethash/consensus.go]: Typo!
Should be 
```
	bigTime := new(big.Int).Set(time)
	bigParentTime := new(big.Int).Set(parent.Time)
```
No?

File: core/types/transaction_signing.go [hunk 2]
```diff
@@ -27,7 +27,12 @@ import (
 	"github.com/ethereum/go-ethereum/params"
 )
 
-var ErrInvalidChainId = errors.New("invalid chaid id for signer")
+var (
+	ErrInvalidChainId = errors.New("invalid chaid id for signer")
+
+	errAbstractSigner     = errors.New("abstract signer")
+	abstractSignerAddress = common.HexToAddress("ffffffffffffffffffffffffffffffffffffff")
```
Comment (holiman) [inline core/types/transaction_signing.go]: You have `ffffffffffffffffffffffffffffffffffffff`, that's only 38 characters. Should be 40, right ? 
https://github.com/ethereum/EIPs/issues/86

File: params/config.go [hunk 3]
```diff
@@ -125,6 +131,7 @@ func (c *ChainConfig) String() string {
 		c.EIP150Block,
 		c.EIP155Block,
 		c.EIP158Block,
+		c.MetropolisBlock,
```
Comment (holiman) [inline params/config.go]: You need to add it to the format-statement aswell

File: core/vm/jump_table.go [hunk 4]
```diff
@@ -47,13 +47,32 @@ type operation struct {
 	// jumps indicates whether operation made a jump. This prevents the program
 	// counter from further incrementing.
 	jumps bool
+	// writes determines whether this a state modifying operation
+	writes bool
```
Comment (holiman) [inline core/vm/jump_table.go]: I'm not sure what's the ultimate intention of `writes` , but you could argue that `CALL` is also a state modifying operation, since if you send ether to a non-existing account, a new account is created in the state. You have not added the `writes`-flag to CALL (and friends) though ?

Comment (obscuren) [inline core/vm/jump_table.go]: Correct, because a call might not always write. This is checked [here](https://github.com/ethereum/go-ethereum/pull/14337/files#diff-8f9edb073f33b63dc5736262579687b8R97). I guess this could have been moved to the metro PR to make things more clear. 

In any case a CALL which does not transfer value (including sub calls) may be regarded as a non-writing operation.

File: core/chain_makers.go [hunk 5]
```diff
@@ -209,15 +209,23 @@ func makeHeader(config *params.ChainConfig, parent *types.Block, state *state.St
 	} else {
 		time = new(big.Int).Add(parent.Time(), big.NewInt(10)) // block time is fixed at 10 seconds
 	}
+	parentHeader := parent.Header()
+	// adjust the parent time
+	parentHeader.Time = new(big.Int).Sub(time, big.NewInt(10))
```
Comment (karalabe) [inline core/chain_makers.go]: You didn't use parentHeader. How didn't the compiler catch this :)) Please drop these 3 lines added here.

File: core/vm/contracts.go [hunk 6]
```diff
@@ -27,15 +28,17 @@ import (
...
+var errBadPrecompileInput = errors.New("bad pre compile input")
+
 // Precompiled contract is the basic interface for native Go contracts. The implementation
 // requires a deterministic gas count based on the input size of the Run method of the
 // contract.
 type PrecompiledContract interface {
-	RequiredGas(inputSize int) uint64 // RequiredPrice calculates the contract gas use
-	Run(input []byte) []byte          // Run runs the precompiled contract
+	RequiredGas(input []byte) uint64  // RequiredPrice calculates the contract gas use
+	Run(input []byte) ([]byte, error) // Run runs the precompiled contract
```
Comment (karalabe) [inline core/vm/contracts.go]: Why do you need to return an error? What issue can bubble up from the VM into the node?

File: core/vm/contracts_test.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package vm
```
Comment (karalabe) [inline core/vm/contracts_test.go]: This is an empty file. Perhaps drop until the actual content is added.

File: core/vm/evm.go [hunk 8]
```diff
@@ -33,7 +33,20 @@ type (
 	GetHashFunc func(uint64) common.Hash
 )
 
-// Context provides the EVM with auxiliary information. Once provided it shouldn't be modified.
+// run runs the given contract and takes care of running precompiles with a fallback to the byte code interpreter.
+func run(evm *EVM, snapshot int, contract *Contract, input []byte) ([]byte, error) {
+	if contract.CodeAddr != nil {
+		precompiledContracts := PrecompiledContracts
```
Comment (karalabe) [inline core/vm/evm.go]: Why copy this map into a local variable?

File: core/types/transaction_signing.go [hunk 9]
```diff
@@ -103,6 +108,17 @@ type Signer interface {
...
+/*
+// WithSignature returns a new transaction with the given signature. This signature
+// needs to be in the [R || S || V] format where V is 0 or 1.
+func (s EIP86Signer) WithSignature(tx *Transaction, sig []byte) (*Transaction, error) {
+}
+
+// Hash returns the hash to be signed by the sender.
+// It does not uniquely identify the transaction.
+func (s EIP86Signer) Hash(tx *Transaction) common.Hash {}
+*/
```
Comment (karalabe) [inline core/types/transaction_signing.go]: Please remove this from the prep PR.

File: core/vm/evm.go [hunk 10]
```diff
@@ -275,10 +294,8 @@ func (evm *EVM) Create(caller ContractRef, code []byte, gas uint64, value *big.I
 	// when we're in homestead this also counts for code storage gas errors.
 	if maxCodeSizeExceeded ||
 		(err != nil && (evm.ChainConfig().IsHomestead(evm.BlockNumber) || err != ErrCodeStoreOutOfGas)) {
+		contract.UseGas(contract.Gas)
```
Comment (karalabe) [inline core/vm/evm.go]: Could you please expand on what's happening here? Why did you change the previous code to this?

File: core/vm/interpreter.go [hunk 11]
```diff
@@ -45,50 +45,60 @@ type Config struct {
 	DisableGasMetering bool
 	// Enable recording of SHA3/keccak preimages
 	EnablePreimageRecording bool
-	// JumpTable contains the EVM instruction table. This
+	// JumpTable contains the in instruction table. This
```
Comment (karalabe) [inline core/vm/interpreter.go]: Accidental rename?

File: core/vm/interpreter.go [hunk 12]
```diff
@@ -45,50 +45,60 @@ type Config struct {
...
 	// may me left uninitialised and will be set the default
 	// table.
 	JumpTable [256]operation
 }
 
 // Interpreter is used to run Ethereum based contracts and will utilise the
-// passed environment to query external sources for state information.
+// passed evmironment to query external sources for state information.
```
Comment (karalabe) [inline core/vm/interpreter.go]: Typo.

File: core/vm/interpreter.go [hunk 13]
```diff
@@ -45,50 +45,60 @@ type Config struct {
...
 	}
 }
 
-// Run loops and evaluates the contract's code with the given input data
-func (evm *Interpreter) Run(contract *Contract, input []byte) (ret []byte, err error) {
-	evm.env.depth++
-	defer func() { evm.env.depth-- }()
+func (in *Interpreter) enforceRestrictions(op OpCode, operation operation, stack *Stack) error {
+	return nil
+}
```
Comment (karalabe) [inline core/vm/interpreter.go]: Plop, what's this?

File: core/vm/interpreter.go [hunk 14]
```diff
@@ -105,35 +115,39 @@ func (evm *Interpreter) Run(contract *Contract, input []byte) (ret []byte, err e
...
+		if err != nil && in.cfg.Debug {
 			// XXX For debugging
 			//fmt.Printf("%04d: %8v    cost = %-8d stack = %-8d ERR = %v\n", pc, op, cost, stack.len(), err)
-			evm.cfg.Tracer.CaptureState(evm.env, pc, op, contract.Gas, cost, mem, stack, contract, evm.env.depth, err)
+			in.cfg.Tracer.CaptureState(in.evm, pc, op, contract.Gas, cost, mem, stack, contract, in.evm.depth, err)
 		}
 	}()
 
-	log.Debug("EVM running contract", "hash", codehash[:])
+	log.Debug("in running contract", "hash", codehash[:])
```
Comment (karalabe) [inline core/vm/interpreter.go]: `in`? Please set to `Interpreter`.

File: core/vm/interpreter.go [hunk 15]
```diff
@@ -105,35 +115,39 @@ func (evm *Interpreter) Run(contract *Contract, input []byte) (ret []byte, err e
...
 	tstart := time.Now()
-	defer log.Debug("EVM finished running contract", "hash", codehash[:], "elapsed", time.Since(tstart))
+	defer log.Debug("in finished running contract", "hash", codehash[:], "elapsed", time.Since(tstart))
```
Comment (karalabe) [inline core/vm/interpreter.go]: `in`? Please set to `Interpreter`.

File: core/vm/interpreter.go [hunk 16]
```diff
@@ -105,35 +115,39 @@ func (evm *Interpreter) Run(contract *Contract, input []byte) (ret []byte, err e
...
 
 	// The Interpreter main run loop (contextual). This loop runs until either an
 	// explicit STOP, RETURN or SELFDESTRUCT is executed, an error occurred during
-	// the execution of one of the operations or until the evm.done is set by
+	// the execution of one of the operations or until the in.done is set by
```
Comment (karalabe) [inline core/vm/interpreter.go]: mass rename typo.

File: core/vm/interpreter.go [hunk 17]
```diff
@@ -194,6 +209,11 @@ func (evm *Interpreter) Run(contract *Contract, input []byte) (ret []byte, err e
 		case !operation.jumps:
 			pc++
 		}
+		// if the operation returned a value make sure that is also set
+		// the last return data.
+		if res != nil {
+			mem.lastReturn = ret
+		}
```
Comment (karalabe) [inline core/vm/interpreter.go]: Just to make sure, this doens't change the current behavior of the EVM, right? What happens if the last opcode in a current contract is CALL?

File: core/vm/jump_table.go [hunk 18]
```diff
@@ -47,13 +47,32 @@ type operation struct {
...
 	// valid is used to check whether the retrieved operation is valid and known
 	valid bool
+	// reverts determined whether the operation reverts state
+	reverts bool
 }
 
-var defaultJumpTable = NewJumpTable()
+var (
+	baseInstructionSet      = NewBaseInstructionSet()
```
Comment (karalabe) [inline core/vm/jump_table.go]: Could we call this `frontierInstructionSet` and `NewFrontierInstructionSet`? We did the same for Frontier/Homesteaddifficulty calculations and also for signers too afaik. Would make the code more uniform.

File: core/vm/jump_table.go [hunk 19]
```diff
@@ -47,13 +47,32 @@ type operation struct {
...
+	homesteadInstructionSet = NewHomesteadInstructionSet()
+)
+
+func NewHomesteadInstructionSet() [256]operation {
```
Comment (karalabe) [inline core/vm/jump_table.go]: dox

File: core/vm/jump_table.go [hunk 20]
```diff
@@ -47,13 +47,32 @@ type operation struct {
...
+		gasCost:       gasDelegateCall,
+		validateStack: makeStackFunc(6, 1),
+		memorySize:    memoryDelegateCall,
+		valid:         true,
+	}
+	return instructionSet
+}
 
-func NewJumpTable() [256]operation {
+func NewBaseInstructionSet() [256]operation {
```
Comment (karalabe) [inline core/vm/jump_table.go]: dox

File: params/config.go [hunk 21]
```diff
@@ -26,28 +26,32 @@ import (
...
+		HomesteadBlock:  MainNetHomesteadBlock,
+		DAOForkBlock:    MainNetDAOForkBlock,
+		DAOForkSupport:  true,
+		EIP150Block:     MainNetHomesteadGasRepriceBlock,
+		EIP150Hash:      MainNetHomesteadGasRepriceHash,
+		EIP155Block:     MainNetSpuriousDragon,
+		EIP158Block:     MainNetSpuriousDragon,
+		MetropolisBlock: MainNetMetropolisBlock,
+
+		Ethash: new(EthashConfig),
```
Comment (karalabe) [inline params/config.go]: I'd recommend dropping all MainNetxxx constants and instead moving them into this struct. It's ambiguous from outside code that there's two way to retrieve the same constants and aso error prone if you add a constant but don't add it to the chain config.

Comment (karalabe) [inline params/config.go]: Similar to the `TestnetChainConfig` below.

File: params/config.go [hunk 22]
```diff
@@ -26,28 +26,32 @@ import (
...
-		Ethash:         new(EthashConfig),
+		ChainId:         big.NewInt(3),
+		HomesteadBlock:  big.NewInt(0),
+		DAOForkBlock:    nil,
+		DAOForkSupport:  true,
+		EIP150Block:     big.NewInt(0),
+		EIP150Hash:      common.HexToHash("0x41941023680923e0fe4d74a34bdac8141f2540e3ae90623718e47d66d1ca4a2d"),
+		EIP155Block:     big.NewInt(10),
+		EIP158Block:     big.NewInt(10),
+		MetropolisBlock: TestNetMetropolisBlock,
```
Comment (karalabe) [inline params/config.go]: I'd recommend using the real value, not an intermediate constant here. It'd be cleaner.

File: params/config.go [hunk 23]
```diff
@@ -26,28 +26,32 @@ import (
...
+
+		Ethash: new(EthashConfig),
 	}
 
 	// RinkebyChainConfig contains the chain parameters to run a node on the Rinkeby test network.
```
Comment (karalabe) [inline params/config.go]: Please add the Metropolis block field to RinkebyConfig too.

File: params/config.go [hunk 24]
```diff
@@ -95,8 +99,10 @@ type ChainConfig struct {
 	EIP150Block *big.Int    `json:"eip150Block,omitempty"` // EIP150 HF block (nil = no fork)
 	EIP150Hash  common.Hash `json:"eip150Hash,omitempty"`  // EIP150 HF hash (fast sync aid)
 
-	EIP155Block *big.Int `json:"eip155Block,omitempty"` // EIP155 HF block
-	EIP158Block *big.Int `json:"eip158Block,omitempty"` // EIP158 HF block
+	EIP155Block *big.Int `json:"eip155Block"` // EIP155 HF block
+	EIP158Block *big.Int `json:"eip158Block"` // EIP158 HF block
```
Comment (karalabe) [inline params/config.go]: Please readd the omitempty tags

File: params/config.go [hunk 25]
```diff
@@ -95,8 +99,10 @@ type ChainConfig struct {
...
+
+	MetropolisBlock *big.Int `json:"metropolisBlock"` // Metropolis switch block (nil = no fork, 0 = alraedy on homestead)
```
Comment (karalabe) [inline params/config.go]: Please add the omitempty tag.

File: params/util.go [hunk 26]
```diff
@@ -38,6 +39,9 @@ var (
 	TestNetSpuriousDragon = big.NewInt(10)
 	MainNetSpuriousDragon = big.NewInt(2675000)
 
-	TestNetChainID = big.NewInt(3) // Testnet default chain ID
-	MainNetChainID = big.NewInt(1) // Mainnet default chain ID
+	TestNetMetropolisBlock = big.NewInt(math.MaxInt64)
+	MainNetMetropolisBlock = big.NewInt(math.MaxInt64)
+
+	TestNetChainID = big.NewInt(3) // Test net default chain ID
+	MainNetChainID = big.NewInt(1) // main net default chain ID
```
Comment (karalabe) [inline params/util.go]: I'd recommend dropping all these fields (,aybe the hash needs to stay and using the already existing ChainConfig constants instead.

File: core/vm/contracts.go (hunk 6 above)
Comment (obscuren) [inline core/vm/contracts.go]: It's required in the metro PR.

File: core/vm/contracts_test.go (hunk 7 above)
Comment (obscuren) [inline core/vm/contracts_test.go]: Meh :)

File: core/vm/evm.go (hunk 8 above)
Comment (obscuren) [inline core/vm/evm.go]: To make my life easier in the follow up PR. I literally split the PR in to two pieces and this was the easiest way.

File: core/vm/interpreter.go (hunk 11 above)
Comment (obscuren) [inline core/vm/interpreter.go]: I guess. Thanks

File: core/vm/interpreter.go (hunk 12 above)
Comment (obscuren) [inline core/vm/interpreter.go]: Fixed in the metro PR

File: core/vm/interpreter.go (hunk 13 above)
Comment (obscuren) [inline core/vm/interpreter.go]: See metro PR. Merely to make my life easier in the follow up.

File: core/vm/interpreter.go (hunk 14 above)
Comment (obscuren) [inline core/vm/interpreter.go]: This, and the previous instance, is probably because of a `s/EVM/in/g` or something like that.

File: core/vm/interpreter.go (hunk 17 above)
Comment (obscuren) [inline core/vm/interpreter.go]: No, this shouldn't change the behaviour of the EVM.

File: core/vm/jump_table.go (hunk 18 above)
Comment (obscuren) [inline core/vm/jump_table.go]: Sure.

**Comment by holiman** [comment]
I'd like to test it a bit in hive, but can't build it as it is now: 
```
# github.com/ethereum/go-ethereum/consensus/ethash
consensus/ethash/consensus.go:318: cannot use time (type uint64) as type *big.Int in argument to new(big.Int).Set
```

File: core/vm/evm.go (hunk 10 above)
Comment (obscuren) [inline core/vm/evm.go]: Because it contains duplicate code. The return statement does the same thing as the return couple lines below.

File: params/util.go (hunk 26 above)
Comment (obscuren) [inline params/util.go]: We can do that. I'll make that change in the metro PR.

**Comment by obscuren** [comment]
@holiman: fixed
